### PR TITLE
Hit rate fix in log unit cache

### PR DESCRIPTION
**Pull request by xnull**
Hit rate fix in log unit cache. Now we collect hit hate and aggregate it by minute

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Review by PavelZaytsev — APPROVED** [review]

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java [hunk 1]
```diff
@@ -55,12 +57,30 @@ public LogUnitServerCache(LogUnitServerConfig config, StreamLog streamLog) {
...
-        MicroMeterUtils.gauge(weightName, dataCache, cache -> cache.stats().evictionWeight());
+                .ifPresent(registry -> {
+                    CaffeineCacheMetrics.monitor(registry, dataCache, "logunit.read_cache");
+
+                    //calc hit rate
+                    MicroMeterUtils.gauge(hitRatioName, dataCache, cache -> {
+                        double miss = getReadCacheCounter(registry, Tag.of("result", "miss")).count();
+                        double hit = getReadCacheCounter(registry, Tag.of("result", "hit")).count();
+
+                        return hit / (hit + miss);
```
Comment (zfrenette) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:68]: Just wondering - If both `miss` and `hit` are 0, it looks like this result will be `NaN`. Is this handled?

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java [hunk 2]
```diff
@@ -55,12 +57,30 @@ public LogUnitServerCache(LogUnitServerConfig config, StreamLog streamLog) {
 
         MeterRegistryProvider
                 .getInstance()
-                .ifPresent(registry -> CaffeineCacheMetrics.monitor(registry, dataCache, "logunit.read_cache"));
-        MicroMeterUtils.gauge(hitRatioName, dataCache, cache -> cache.stats().hitRate());
```
Comment (SravanthiAshokKumar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:59]: Just for my understanding : what was the bug with using the hitRate() stat?

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:59]: The rate is calculated over the life time of the cache object, but the reporting interval is per-minute. So if the cache has a high hit rate for a long period of time then a sudden drops in the hit rate wouldn't be reflected clearly. For example, if you have 100 requests per minute and the first 5 minutes the cache has a 100% hit rate, then on the 6th minute you have a 0 hit rate the calculated average would be 500 (hitRate) /600(all requests) ~ 83.3%, this can be misleading, ideally we want the reported hit rate on the 6th minute to be 0%.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java (hunk 1 above)
Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:68]: I would suggest not calculating these manually, but instead use the `CacheStats` object which implements helper methods that account for these edge cases. For example, if  `oldVal = cache.states()` and `newVal = cache.states()`, then you can do `newVal.minutes(oldVal).hitRate()`

Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: > Just wondering - If both `miss` and `hit` are 0, it looks like this result will be `NaN`. Is this handled?

I have fixed that issue, thank you for the hint

Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: > I would suggest not calculating these manually, but instead use the `CacheStats` object which implements helper methods that account for these edge cases. For example, if `oldVal = cache.states()` and `newVal = cache.states()`, then you can do `newVal.minus(oldVal).hitRate()`
> 
> Note that `cache.states()` takes a snapshot of the stats at that particular time.

In general I agree, but:
 - I don't like the idea of keeping stats snapshot (and it looks like the link to it needs to be mutable)
 - calculating the diff between parsentages between two shapshots seems tedious
 - in my solution we calculate the value out of the numbers we get right away. The calculation itself it pretty simple and straighforward
 - Using my approach we can calculate other numbers also, if needed

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: I don't understand the points. Can you rephrase? (preferably in a more succinct and objective way)

Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: It's easy and simple just to take two numbers and divide one by another rather than doing something else

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        LogUnitServerCache cache = new LogUnitServerCache(config, streamLog);
+
+        for (int addr = 1; addr < 101; addr++) {
+            LogData logData = generateLogData(addr);
+            streamLog.append(addr, logData);
+        }
+
+        for (int i = 0; i < 500; i++) {
+            cache.get(rnd.nextInt(150));
+        }
```
Comment (zfrenette) [inline infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java:52]: I might have misunderstood something, it seems like there's a bug in the computation of the statistics, the test, or both. I modified the test here to do `cache.get(i)` for `i` between [1, 100] inclusive in order to remove the randomization and have a deterministic result.

Wouldn't we expect the hit ratio to be 1 since the cache contains all of the addresses between [1, 100]? But expected is producing 0 and the test is not failing.

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            Assertions.fail("Metrics not initialized");
+        }
+
+        missAndHit.ifPresent((miss, hit) -> {
+            //compare with hit ratio
+            double expected = hit.count() / (hit.count() + miss.count());
+            if (cache.hitRatio().isPresent()) {
+                assertEquals(expected, cache.hitRatio().get().value());
+            } else {
+                Assertions.fail("Hit ration not present");
```
Comment (zfrenette) [inline infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java:65]: nit: ratio.

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java (hunk 3 above)
Comment (zfrenette) [reply infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java:52]: I see my misunderstanding. The append doesn't actually populate the cache, so the first get() for an address is a miss.

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Random;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+class LogUnitServerCacheTest {
+    private final Random rnd = new Random();
+
+    @Test
+    public void testMetrics() {
```
Comment (zfrenette) [inline infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java:27]: Can we add a test without any accesses to exercise the `NaN` case?

**Review by zfrenette — APPROVED** [review]
Can we add a few deterministic tests (hitRatio of 1, 0...) so that detecting failures from bugs in the future can be done more reliably?

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java (hunk 4 above)
Comment (xnull) [reply infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java:75]: done

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java (hunk 5 above)
Comment (xnull) [reply infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java]: done

**Comment by xnull** [comment]
> Can we add a few deterministic tests (hitRatio of 1, 0, ...) so that detecting failures from bugs in the future can be done more reliably?

I just added the test for the case of zero hits and 1.0 hit ratio. If you know any other scenarios to cover, just let me know, I'd be happy to add them

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java (hunk 1 above)
Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: Hey Maithem, 
I'he got some logs [here](https://gist.github.com/xnull/2808aa48c27b66e6ccec87cfeeee7af9)
As you can see, the `miss` metric has been aggregated by minute - the value going up and down each minute, which means it really has been aggregated by minute.
The same true for `hit` metric, just by definition.

So, the only thing that I need to do is to divide one number by another (miss vs hit) the same way how it's done in the stats class and I will get hit rate aggregated by minute too.

File: infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        LogUnitServerConfig config = LogUnitServerConfig.builder()
+                .cacheSizeHeapRatio(cacheSizeHeapRatio)
+                .maxCacheSize(maxCacheSize)
+                .memoryMode(true)
+                .noSync(true)
+                .build();
+        return new LogUnitServerCache(config, streamLog);
+    }
+
+    private static void init_metrics() {
```
[the remark this answers is not in the document]
Comment (xnull) [reply infrastructure/src/test/java/org/corfudb/infrastructure/LogUnitServerCacheTest.java]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java (hunk 1 above)
Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: To get non cumulative states there are two ways. Either have a counter that gets reset per-poll period, or you subtract two cumulative counters to get the difference. Instead of showing logs can you write a test case that demonstrates correct per-poll period stats?

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]:
```java

Optional<FunctionCounter> miss() {
        return MeterRegistryProvider.getInstance()
                .map(registry -> registry.get("cache.gets").tags(missTags).functionCounter());
    }
```

This code snippet will generate a counter with a weak-reference, so if the object is getting gc'd then the counter will reset, but it wouldnt be reporting correct statistics.

**Review by chetangudisagar — APPROVED** [review]

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java (hunk 1 above)
Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: > ldnt be reporting correct statistics.

I can do that, but that's more of verifying of how Micrometer works underlying. We know that it works in a correct way, otherwise any metric will be affected and the entire thing wouldn't work as expected.

Anyway, let me write a test for that, even though it's a bit complex and time consuming.

Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: > ```java
> Optional<FunctionCounter> miss() {
>         return MeterRegistryProvider.getInstance()
>                 .map(registry -> registry.get("cache.gets").tags(missTags).functionCounter());
>     }
> ```
> 
> This code snippet will generate a counter with a weak-reference, so if the object is getting gc'd then the counter will reset, but it wouldnt be reporting correct statistics.

I'm not sure if that gonna be a problem at all. In our hitRatio metrics code we get a  miss and hit counter and use it immediately, I don't see how there could a problem. Even if your scenario happens once in a very rare probability time, we just get a 1.0 at that minute which is not a big of a deal. 
And the same will happen with your approach too, anyway

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: > And the same will happen with your approach too, anyway

False. I think they gauges should have strong references so that issue wouldn't happen.

Comment (zfrenette) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: @xnull In a test, it doesn't seem like the logunit_cache_hit_ratio metric changes to 1 in a minute with no activity. Rather, it keeps the same value that it had in previous minutes.

Comment (zfrenette) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java]: From three consecutive minutes (the later two have no activity):

logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843
logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843
logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843
